semver.py

```diff
@@ -288,6 +288,29 @@ def parse(version):
         """
         return parse_version_info(version)
 
+    def replace(self, **parts):
+        """Replace one or more parts of a version and return a new
+          :class:`semver.VersionInfo` object, but leave self untouched
+
+        :param dict parts: the parts to be updated. Valid keys are:
+          ``major``, ``minor``, ``patch``, ``prerelease``, or ``build``
+        :return: the new :class:`semver.VersionInfo` object with the changed
+          parts
+        :raises: KeyError, if ``parts`` contains invalid keys
+        """
+        d1 = self._asdict()
+        d1.update(parts)
+        try:
+            return VersionInfo(**d1)
+        except ValueError:
+            parts.pop('major', None)
+            parts.pop('minor', None)
+            parts.pop('patch', None)
+            parts.pop('prerelease', None)
+            parts.pop('build', None)
+            keys = ", ".join([k for k in parts.keys()])
```

Comment:
Wouldn't a set-diff be more expressive? Not sure if this works as-is but I think: `parts.keys() - d1.keys()`

Oh and I'd use a better name than `d1`

Reply:
> Wouldn't a set-diff be more expressive?

Good idea! See below.

> Not sure if this works as-is but I think: `parts.keys() - d1.keys()`

With sets, this won't be necessary.

> Oh and I'd use a better name than d1

Yes. :+1:  

Hope this is better:

```python
version = self._asdict()
version.update(parts)
try:
    return VersionInfo(**version)
except ValueError:
    parts.pop('major', None)
    parts.pop('minor', None)
    parts.pop('patch', None)
    parts.pop('prerelease', None)
    parts.pop('build', None)
    keys = ", ".join(set(parts) - set(version))
    raise KeyError("Invalid key(s): %s" % keys)
```

Reply:
It still suffers from a wrong conversion of ValueError->KeyError. I'd check for keyerror first (via set diff) and remove the try completely